// material-ui-next-build/MobileStepper/MobileStepper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.styleSheet = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _jssThemeReactor = require('jss-theme-reactor');

var _withStyles = require('../styles/withStyles');

var _withStyles2 = _interopRequireDefault(_withStyles);

var _Paper = require('../Paper');

var _Paper2 = _interopRequireDefault(_Paper);

var _Button = require('../Button');

var _Button2 = _interopRequireDefault(_Button);

var _helpers = require('../utils/helpers');

var _keyboardArrowLeft = require('../svg-icons/keyboard-arrow-left');

var _keyboardArrowLeft2 = _interopRequireDefault(_keyboardArrowLeft);

var _keyboardArrowRight = require('../svg-icons/keyboard-arrow-right');

var _keyboardArrowRight2 = _interopRequireDefault(_keyboardArrowRight);

var _Progress = require('../Progress');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var styleSheet = exports.styleSheet = (0, _jssThemeReactor.createStyleSheet)('MuiMobileStepper', function (theme) {
  return {
    root: {
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center',
      background: theme.palette.background.default,
      padding: theme.spacing.unit
    },
    positionBottom: {
      position: 'fixed',
      bottom: 0,
      left: 0,
      right: 0,
      zIndex: theme.zIndex.mobileStepper
    },
    positionTop: {
      position: 'fixed',
      top: 0,
      left: 0,
      right: 0,
      zIndex: theme.zIndex.mobileStepper
    },
    positionStatic: {},
    button: {},
    dots: {
      display: 'flex',
      flexDirection: 'row'
    },
    dot: {
      backgroundColor: theme.palette.action.disabled,
      borderRadius: '50%',
      width: theme.spacing.unit,
      height: theme.spacing.unit,
      margin: '0 2px'
    },
    dotActive: {
      backgroundColor: theme.palette.primary[500]
    },
    progress: {
      width: '50%'
    }
  };
}); //  weak

var _ref = _react2.default.createElement(_keyboardArrowLeft2.default, null);

var _ref2 = _react2.default.createElement(_keyboardArrowRight2.default, null);

function MobileStepper(props) {
  var activeStep = props.activeStep,
      backButtonText = props.backButtonText,
      classes = props.classes,
      classNameProp = props.className,
      disableBack = props.disableBack,
      disableNext = props.disableNext,
      position = props.position,
      type = props.type,
      nextButtonText = props.nextButtonText,
      onBack = props.onBack,
      onNext = props.onNext,
      steps = props.steps,
      other = (0, _objectWithoutProperties3.default)(props, ['activeStep', 'backButtonText', 'classes', 'className', 'disableBack', 'disableNext', 'position', 'type', 'nextButtonText', 'onBack', 'onNext', 'steps']);


  var className = (0, _classnames2.default)(classes.root, classes['position' + (0, _helpers.capitalizeFirstLetter)(position)], classNameProp);

  return _react2.default.createElement(
    _Paper2.default,
    (0, _extends3.default)({ square: true, elevation: 0, className: className }, other),
    _react2.default.createElement(
      _Button2.default,
      { className: classes.button, onClick: onBack, disabled: disableBack },
      _ref,
      backButtonText
    ),
    type === 'dots' && _react2.default.createElement(
      'div',
      { className: classes.dots },
      [].concat((0, _toConsumableArray3.default)(new Array(steps))).map(function (_, step) {
        var dotClassName = (0, _classnames2.default)((0, _defineProperty3.default)({}, classes.dotActive, step === activeStep), classes.dot);
        // eslint-disable-next-line react/no-array-index-key
        return _react2.default.createElement('div', { key: step, className: dotClassName });
      })
    ),
    type === 'progress' && _react2.default.createElement(
      'div',
      { className: classes.progress },
      _react2.default.createElement(_Progress.LinearProgress, { mode: 'determinate', value: Math.ceil(activeStep / (steps - 1) * 100) })
    ),
    _react2.default.createElement(
      _Button2.default,
      { className: classes.button, onClick: onNext, disabled: disableNext },
      nextButtonText,
      _ref2
    )
  );
}

MobileStepper.propTypes = process.env.NODE_ENV !== "production" ? {
  /**
   * Set the active step (zero based index).
   * Defines which dot is highlighted when the type is 'dots'.
   */
  activeStep: _propTypes2.default.number,
  /**
   * Set the text that appears for the back button.
   */
  backButtonText: _propTypes2.default.node,
  /**
   * Useful to extend the style applied to components.
   */
  classes: _propTypes2.default.object.isRequired,
  /**
   * @ignore
   */
  className: _propTypes2.default.string,
  /**
   * Set to true to disable the back button.
   */
  disableBack: _propTypes2.default.bool,
  /**
   * Set to true to disable the next button.
   */
  disableNext: _propTypes2.default.bool,
  /**
   * Set the text that appears for the next button.
   */
  nextButtonText: _propTypes2.default.node,
  /**
   * Passed into the onClick prop of the Back button.
   */
  onBack: _propTypes2.default.func.isRequired,
  /**
   * Passed into the onClick prop of the Next button.
   */
  onNext: _propTypes2.default.func.isRequired,
  /**
   * Set the text that appears for the next button.
   */
  position: _propTypes2.default.oneOf(['bottom', 'top', 'static']),
  /**
   * The total steps.
   */
  steps: _propTypes2.default.number.isRequired,
  /**
   * The type of mobile stepper to use.
   */
  type: _propTypes2.default.oneOf(['text', 'dots', 'progress'])
} : {};

MobileStepper.defaultProps = {
  activeStep: 0,
  backButtonText: 'Back',
  disableBack: false,
  disableNext: false,
  nextButtonText: 'Next',
  position: 'bottom',
  type: 'dots'
};

exports.default = (0, _withStyles2.default)(styleSheet)(MobileStepper);
